### DISCUSSION TITLE
Linux DRM run Avalonia App for each monitor

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#14881] [Q&A] [answered]
In pr #14878 I allowed Avalonia to select the render output monitor (Connector_Type, Connector_Id). The next step is to allow run Avalonia App for each monitor, but when I run the second app, it crashes. From the debugging I did, it failed in the following line

https://github.com/AvaloniaUI/Avalonia/blob/fbae909ec86f70471378425f0f00c8ede3be15a1/src/Linux/Avalonia.LinuxFramebuffer/Output/DrmOutput.cs#L212

The error result is EACCES(13)  Permission denied.

Do you have any suggestions on how to resolve this?

Avalonia Version: **fbae909ec86f70471378425f0f00c8ede3be15a1**
OS: **`Alpine Extended 3.18.0 x86_64`**
HW:  
 - CPU: `Intel(R) Celeron(R) CPU  J1900  @ 1.99GHz`,  
 - VGA: `Intel(R) VLV Mobile/Desktop Graphics Controller`

**Comment by kekekeks** [comment, marked as answer]
>Do you have any suggestions on how to resolve this?

There can be only one DRM master for a particular tty. You can try pass the DRM fd via `sendfd` or just run one app for multiple outputs.

We can also add some API to use multiple DRM outputs from one application, but input management will be rather tricky

**Reply by workgroupengineering** [reply]
Hi,

 I try to implement [senfd](https://gist.github.com/jim3ma/93c41ac2bdf347579524b81064b5b03d) but not i have success.

> just run one app for multiple outputs.

How to do it? My app has different output on each monitor.

**Reply by workgroupengineering** [reply]
After some trouble, I success to implement it. Before make a PR i should make code refactoring.
I thought of a similar solution:

```mermaid
---
title: DRMOutput Refactoring
---
classDiagram
    
    MasterDrmOutput <|-- DrmOutputBase
    ClientDrmOutput <|-- DrmOutputBase
    DrmOutput <|-- DrmOutputBase
    MasterOutputOptions <|-- DrmOutputOptions
    MasterOutputOptions : +String EndPoint
    MasterOutputOptions : +int MaxClients
    ClientOutputOptions <|-- DrmOutputOptions
    ClientOutputOptions : +String EndPoint
    class DrmOutputBase{
        DrmOutputBase(DrmCard card, DrmResources resources, DrmConnector connector, DrmModeInfo modeInfo)        
    }
    class MasterDrmOutput{
        MasterDrmOutput(string path = null, bool connectorsForceProbe = false, MasterOutputOptions options = null)      
    }
    class ClientDrmOutput{
        ClientDrmOutput(string path = null, bool connectorsForceProbe = false, ClientOutputOptions options = null)      
    }
    class DrmOutputOptions{
        double Scaling
        bool EnableInitialBufferSwapping
        Color InitialBufferSwappingColor
        PixelSize? VideoMode
        DrmConnectorType? ConnectorType 
        uint? ConnectorType_Id
    }

```

What do you think?

**Reply by kekekeks** [reply]
I think drm master management should happen outside of Avalonia code for now, since unix socket server for managing DRM clients is certainly out of scope of Avalonia. Especially since there are other providers of DRM lease.
 Extracting libdrm bindings into a separate library and having a separate lib on top of it would probably the right way of doing it.  

On Avalonia side we probably should just accept the card fd in DrmOutput and make sure that we can work with fds received via https://wayland.app/protocols/drm-lease-v1 protocol

**Reply by workgroupengineering** [reply]
DRM is independent of Wayland, right?

**Reply by kekekeks** [reply]
Yes, the protocol simply passes a fd

**Reply by workgroupengineering** [reply]
If I understand correctly, you suggest adding the  parameter `int?  masterCrtcId` to the DrmOutput constructor.
For small apps, all this management seems excessive to me. How could this be remedied?

**Reply by kekekeks** [reply]
Why do we need to pass crtc id? Just introduce an overload that accepts the card `fd` instead of path.
A "leased" DRM fd isn't really distinguishable from a full one from the app point of view, the only difference is

>Objects not controlled by a drm_master cannot be modified through the various state manipulating ioctls, and any state reported back to user space will be edited to make them appear idle and/or unusable. For instance, connectors always report 'disconnected', while encoders report no possible crtcs or clones.

**Reply by kekekeks** [reply]
> For small apps, all this management seems excessive to me. How could this be remedied?

DRM lease is an advanced usage scenario that is only required for multiprocess scenarios. If one wants to have multiprocess DRM sharing, there will be some manual setup.
